File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:

        counts = Counter(nums)
        buckets = [[] for _ in range(len(nums) + 1)]

        for num, count in counts.items():
            buckets[count].append(num)
        
        total_len = 0
        res = []
        for i in range(len(nums), -1, -1):
            res.extend(buckets[i])
            total_len += len(buckets[i])
            if total_len == k:
                return res

            

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();

        vector<vector<int>> buckets(n + 1);

        unordered_map<int, int> counts;

        for (auto num : nums){
            counts[num]++;
        }

        for (auto key_value : counts){
            int number = key_value.first;
            int count = key_value.second;
            buckets[count].push_back(number);
        }

        std::reverse(buckets.begin(), buckets.end());

        vector<int> result; 
        int i = 0;
        while (result.size() < k) {
            vector<int> current_bucket = buckets[i];
            result.insert(result.end(), current_bucket.begin(), current_bucket.end());
            i++;
        }

        return result;
        
        
    }
};